lib/charms/sdcore_webui_k8s/v0/sdcore_config.py

```diff
@@ -0,0 +1,343 @@
+    ) -> Optional[Dict[str, str]]:
+        """Get relation data for the remote application.
+
+        Args:
+            relation: Juju relation object (optional).
+
+        Returns:
+            Dict: Relation data for the remote application
+            or None if the relation data is invalid.
+        """
+        relation = relation or self.model.get_relation(self.relation_name)
+
+        if not relation:
+            logger.error("No relation: %s", self.relation_name)
+            return None
+
+        if not relation.app:
+            logger.warning("No remote application in relation: %s", self.relation_name)
+            return None
+
+        remote_app_relation_data = dict(relation.data[relation.app])
+
+        if not data_is_valid(remote_app_relation_data):
+            logger.error("Invalid relation data: %s", remote_app_relation_data)
+            return None
+
+        return remote_app_relation_data
+
+
+class SdcoreConfigProvides(Object):
+    """Class to be instantiated by the charm providing the SD-Core Webui URL."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Init."""
+        super().__init__(charm, relation_name)
+        self.relation_name = relation_name
+        self.charm = charm
+
+    def set_webui_url(self, webui_url: str, relation_id: int) -> None:
+        """Set the address of the Webui GRPC endpoint.
+
+        Args:
+            webui_url (str): Webui GRPC service address.
+            relation_id (int): Relation ID.
+
+        Returns:
+            None
+        """
+        if not self.charm.unit.is_leader():
+            raise RuntimeError("Unit must be leader to set application relation data.")
+
+        if not data_is_valid(data={"webui_url": webui_url}):
+            raise ValueError(f"Invalid url: {webui_url}")
+
+        relation = self.model.get_relation(
+            relation_name=self.relation_name, relation_id=relation_id
+        )
+
+        if not relation:
+            raise RuntimeError(f"Relation {self.relation_name} not created yet.")
+
+        if relation not in self.model.relations[self.relation_name]:
+            raise RuntimeError(f"Relation {self.relation_name} not created yet.")
+
+        relation.data[self.charm.app].update({"webui_url": webui_url})
+
+    def set_webui_url_in_all_relations(self, webui_url: str) -> None:
+        """Set Webui URL in applications for all applications.
+
+        Args:
+            webui_url (str): Webui GRPC service address
+
+        Returns:
+            None
+        """
+        if not self.charm.unit.is_leader():
+            raise RuntimeError("Unit must be leader to set application relation data.")
+
+        if not data_is_valid(data={"webui_url": webui_url}):
+            raise ValueError(f"Invalid url: {webui_url}")
+
+        relations = self.model.relations[self.relation_name]
+
+        if not relations:
+            raise RuntimeError(f"Relation {self.relation_name} not created yet.")
```

Comment:
Is this really an error case ? 
I am afraid we would run into a case in which we deploy the webui but the relations are not created yet and we call a `configure` method that calls `set_webui_url_in_all_relations`

Reply:
In the configure method, this method should run after we make sure that sdcore_config relation is set up. We always perform the preliminary checks in the configure methods and I think this would not be a problem

Reply:
The problem I see is that this adds a precondition to the public method and also adds a state

Reply:
Done